src/main/java/com/google/sps/servlets/GroupsServlet.java

```diff
@@ -96,4 +124,28 @@ private String createGroup(String name, String description, String image){
 
     return GROUP_ID_STRING;
   }
+
+  /**
+   * Adds user with email to group with groupId. If the user is not in datastore, a new user is created.
+   */
+  private void addUserToGroup(String groupId, String email) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    try {
+      Entity user = datastore.get(KeyFactory.createKey(USER_KIND, email));
+
+      HashSet<Long> groups = (HashSet<Long>) user.getProperty(GROUPS_KEY);
+      groups.add(Long.valueOf(groupId));
+      user.setProperty(GROUPS_KEY, groups);
+
+      datastore.put(user);
+    } catch (EntityNotFoundException newUser) {
+      Entity user = new Entity(USER_KIND, email);
+
+      user.setProperty(USER_EMAIL_PROPERTY, email);
+      user.setProperty(GROUPS_KEY, new HashSet<Long>());
```

Comment:
I'm not entirely sure on the flow you were going for (so correct me if I'm wrong), but would it be possible to create a new user if they don't already exist, and add them to the group in this same method? for example, on the UI there's a button called "JOIN" for each group. if that  button is pressed, would there even be a possibility that the user does not exist in datastore?